src/hardware/vga_draw.cpp

```diff
@@ -1486,255 +1491,411 @@ static std::pair<Fraction, Fraction> determine_pixel_size(const uint32_t htotal,
+			vert.blanking_start |= (vga.s3.ex_ver_overflow & 0x4) << 8;
+			vert.retrace_start |= ((vga.crtc.overflow & 0x80) << 2);
+			vert.retrace_start |= (vga.s3.ex_ver_overflow & 0x10) << 6;
+			vert.blanking_end = vga.crtc.end_vertical_blanking & 0x7f;
 		} else { // EGA
-			vbend = vga.crtc.end_vertical_blanking & 0x1f;
+			vert.blanking_end = vga.crtc.end_vertical_blanking & 0x1f;
+		}
+
+		horiz.total += 2;
+		vert.total += 2;
+		horiz.display_end += 1;
+		vert.display_end += 1;
+
+		horiz.blanking_end = horiz.blanking_start +
+		                     ((horiz.blanking_end - horiz.blanking_start) &
+		                      0x3F);
+		horiz.retrace_end = vga.crtc.end_horizontal_retrace & 0x1f;
+		horiz.retrace_end = (horiz.retrace_end - horiz.retrace_start) & 0x1f;
+
+		if (!horiz.retrace_end) {
+			horiz.retrace_end = horiz.retrace_start + 0x1f + 1;
+		} else {
+			horiz.retrace_end = horiz.retrace_start + horiz.retrace_end;
 		}
-		htotal += 2;
-		vtotal += 2;
-		hdend += 1;
-		vdend += 1;
-
-		hbend = hbstart + ((hbend - hbstart) & 0x3F);
-		hrend = vga.crtc.end_horizontal_retrace & 0x1f;
-		hrend = (hrend - hrstart) & 0x1f;
-		
-		if ( !hrend ) hrend = hrstart + 0x1f + 1;
-		else hrend = hrstart + hrend;
-
-		vrend = vga.crtc.vertical_retrace_end & 0xF;
-		vrend = ( vrend - vrstart)&0xF;
-		
-		if ( !vrend) vrend = vrstart + 0xf + 1;
-		else vrend = vrstart + vrend;
-
-		// Special case vbstart==0:
-		// Most graphics cards agree that lines zero to vbend are
-		// blanked. ET4000 doesn't blank at all if vbstart==vbend.
-		// ET3000 blanks lines 1 to vbend (255/6 lines).
-		if (vbstart != 0) {
-			vbstart += 1;
-			vbend = (vbend - vbstart) & 0x7f;
-			if ( !vbend) vbend = vbstart + 0x7f + 1;
-			else vbend = vbstart + vbend;
+
+		vert.retrace_end = vga.crtc.vertical_retrace_end & 0xF;
+		vert.retrace_end = (vert.retrace_end - vert.retrace_start) & 0xF;
+
+		if (!vert.retrace_end) {
+			vert.retrace_end = vert.retrace_start + 0xf + 1;
+		} else {
+			vert.retrace_end = vert.retrace_start + vert.retrace_end;
 		}
-		vbend++;
-			
+
+		// Special case for vert.blanking_start == 0:
+		// Most graphics cards agree that lines zero to vertical blanking end
+		// are blanked.
+		// Tsend ET4000 doesn't blank at all if vert.blanking_start == vert.blanking_end.
+		// ET3000 blanks lines 1 to vert.blanking_end (255/6 lines).
+		if (vert.blanking_start != 0) {
+			vert.blanking_start += 1;
+			vert.blanking_end = (vert.blanking_end - vert.blanking_start) & 0x7f;
+			if (!vert.blanking_end) {
+				vert.blanking_end = vert.blanking_start + 0x7f + 1;
+			} else {
+				vert.blanking_end = vert.blanking_start + vert.blanking_end;
+			}
+		}
+		vert.blanking_end++;
+
 		if (svga.get_clock) {
 			clock = svga.get_clock();
 		} else {
 			switch ((vga.misc_output >> 2) & 3) {
-			case 0:	
-				clock = (machine==MCH_EGA) ? 14318180 : 25175000;
+			case 0:
+				clock = (machine == MCH_EGA) ? CgaPixelClockHz : 25175000;
```

Comment:
can the 2517500 be de-magic-numbered as well?

Reply:
Sure, if you find out what it stands for and tell me 😄

Reply:
You're the VGA expert! :grin:

Looks like the clock for 640x(350/400/480) modes? http://tinyvga.com/vga-timing

Reply:
VGA 640x400 @ 70Hz pixel clock, seems like it.

Source: http://tinyvga.com/vga-timing/640x400@70Hz

<img width="585" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/f272d132-8a46-4843-90f0-833302ab554f">